libs/execution/include/hpx/parallel/executors/parallel_executor.hpp

```diff
@@ -172,12 +176,19 @@ namespace hpx { namespace parallel { namespace execution
         > >
         bulk_async_execute(F && f, S const& shape, Ts &&... ts) const
         {
-            // lazily initialize once
-            static std::size_t global_num_tasks =
-                (std::min)(std::size_t(128), hpx::get_os_thread_count());
-
-            std::size_t num_tasks =
-                (num_tasks_ == std::size_t(-1)) ? global_num_tasks : num_tasks_;
+            std::size_t num_tasks;
+            if (num_tasks_ == std::size_t(-1))
+            {
+                auto pool = threads::get_self_ptr() ?
+                    threads::get_self_id()->get_scheduler_base()->get_parent_pool() :
+                    &hpx::threads::get_thread_manager().default_pool();
+                num_tasks =
+                    (std::min)(std::size_t(128), pool->get_os_thread_count());
+            }
+            else
+            {
+                num_tasks = num_tasks_;
+            }
```

Comment:
I'd suggest to use a `static thread_local` instead. Otherwise this looks very racy (might be benign, though)!

Reply:
Not sure I follow. I wanted to completely avoid `static` variables, so these are local. A `static thread_local` in principle has the same problem as just `static`, although it should be *much* less likely to fail. Which part looks racy to you? `num_tasks_` could potentially be changed since it's public (it's a struct) but I would think that's usually only set in the constructor. We can also make that private.

Reply:
Well, `static thread_local` wouldn't have the same problem. But I agree that getting rid of that static altogether is a Good Thing (tm). We could initialize things in the constructor now, however. 

>  We can also make that private.

Yes, that's a good idea.

Reply:
Updated to intialize the `num_tasks_` in the constructor (it was already private). Also clang-formatted the file, so the diff looks a bit bigger...